### PR TITLE
[next] Change query params type to use string or string[] instead of any

### PR DESCRIPTION
**Pull request by brikou**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by Dru89** [comment]
Hi @brikou.  Thanks for submitting a fix.  It looks like you changed a lot more than just updating the query params to use `string | string[]` instead of any.  

A lot of it looks like it might just be cosmetic changes that make it hard to see what the actual code difference is between the two, though.  Along with fixing the failures in the build, could you also reduce your change down to the actual fix?

I'm not opposed to style changes (though I'm not sure if there's a dedicated style guide for DefinitelyTyped projects) but I'd like to make that at least a separate commit to make reviewing this easier. 😄

**Comment by brikou** [comment]
Yes sorry about that, I will update PR to reduce noise about coding style. As a side note, do you have any idea on how to had getInitialProps signature (for pages)?

**Comment by brikou** [comment]
I've just updated my PR so you can easily review it :+1: ping @Dru89

**Comment by brikou** [comment]
I just removed .prettierrc file as not allowed, ping @Dru89

File: types/next/router.d.ts [hunk 1]
```diff
@@ -14,17 +14,27 @@ export interface SingletonRouter {
     ready(cb: RouterCallback): void;
 
     // router properties
-    readonly components: { [key: string]: { Component: React.ComponentType<any>, err: any } };
+    readonly components: {
+        [key: string]: { Component: React.ComponentType<any>; err: any };
```
Comment (Dru89) [inline types/next/router.d.ts]: You have a semicolon after `<any>;` and I think you want a comma.

File: types/next/router.d.ts [hunk 2]
```diff
@@ -14,17 +14,27 @@ export interface SingletonRouter {
...
+    };
     readonly pathname: string;
     readonly route: string;
     readonly asPath?: string;
-    readonly query?: { [key: string]: any };
+    readonly query?: { [key: string]: string | string[] };
```
Comment (Dru89) [inline types/next/router.d.ts]: I'm guessing this was the main change?

**Review by dru89 — COMMENTED** [review]
It looks like everything is still run through `prettier`.  Can you undo those?  Is the change just what I mentioned in the inline comment (in a couple of places in that file)?  

Also, a small concern I have about this type change is that Node's `querystring` (which is what `next` uses to parse) doesn't actually restrict query values to being `string | string[]`, and instead will take pretty much any value (number, boolean, etc.) so I worry that this might actually be too restrictive.  You can see the documentation on how querystring explicitly allows these other values: https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options

> It serializes the following types of values passed in `obj`: `<string> | <number> | <boolean> | <string[]> | <number[]> | <boolean[]>` Any other input values will be coerced to empty strings.

File: types/next/router.d.ts (hunk 2 above)
Comment (brikou) [reply types/next/router.d.ts]: Yes it is the main change :)

File: types/next/router.d.ts (hunk 1 above)
Comment (brikou) [reply types/next/router.d.ts]: That is the behaviour of prettier and it is correct, you can read more here https://github.com/prettier/prettier/issues/1944

**Comment by brikou** [comment]
I've updated query def according to your comment. Thanks for your time and your great work. Please reconsider prettier formatting I really like it because it is a no brainer. :D

As a side note (not the right PR to talk about it) I found a solution to handle Page special component with getInitialProps, it looks like a hack but work, can you give me your feedback about it?

```ts
// react.d.ts

import "react";
import { ServerRequest, ServerResponse } from "http";

interface IContext {
    asPath: string;
    err?: Error;
    jsonPageRes?: Response;
    pathname: string;
    query: string | string[];
    req?: ServerRequest;
    res?: ServerResponse;
}

// a hack to augment react interfaces
declare module "react" {
    interface StatelessComponent<P = {}> {
        getInitialProps?(context: IContext): Promise<P>;
    }

    interface ComponentClass<P = {}> {
        getInitialProps?(context: IContext): Promise<P>;
    }
}
```

example usage:

```tsx
import * as React from "react";

interface IProps {
    pathname: string;
}

export const Bar: React.SFC<IProps> = ({ pathname }) => (
    <div>
        <span>{pathname}</span>
    </div>
);

Bar.getInitialProps = async (context) => {
    return { pathname: context.pathname };
};

export const Foo: React.SFC<IProps> = ({ pathname }) => (
    <div>
        <span>{pathname}</span>
    </div>
);

Foo.getInitialProps = async () => {
    return { pathname: "/tmp" };
};

export const Baz: React.ComponentClass<IProps> = class extends React.Component<
    IProps
> {
    public render() {
        return (
            <div>
                <span>{this.props.pathname}</span>
            </div>
        );
    }
};

Baz.getInitialProps = async (context) => {
    return { pathname: context.pathname };
};
```
 Would be nice to have your feedback ;)

Cheer

**Comment by brikou** [comment]
ping @Dru89

**Comment by minestarks** [comment]
@Dru89 if you don't have any remaining concerns I'll merge. 

Change LGTM but @brikou it was a bit hard to review with the formatting changes! Tip: review commit-by-commit.

**Pull request merged**